consensus/istanbul/core/core.go

```diff
@@ -263,52 +272,44 @@ func (c *core) startNewRound(round *big.Int) {
 	c.updateRoundState(newView, c.valSet, roundChange)
 	// Calculate new proposer
 	c.valSet.CalcProposer(lastProposer, newView.Round.Uint64())
-	c.waitingForRoundChange = false
+	c.waitingForNewRound = false
 	c.setState(StateAcceptRequest)
 	if roundChange && c.isProposer() && c.current != nil {
-		// If it is locked, propose the old proposal
-		// If we have pending request, propose pending request
-		if c.current.IsHashLocked() {
-			r := &istanbul.Request{
-				Proposal: c.current.Proposal(), //c.current.Proposal would be the locked proposal by previous proposer, see updateRoundState
+		// Start with pending request
+		request := c.current.pendingRequest
+		// Search for a valid request in round change messages.
+		// The round change certificate should be generated such that it is consistent in it's proposed subject.
+		for _, message := range roundChangeCertificate.RoundChangeMessages {
+			var roundChangeMsg *istanbul.RoundChange
+			if err := message.Decode(&roundChangeMsg); err != nil {
+				logger.Error("Failed to decode ROUND CHANGE in certificate. Skipping to next ROUND CHANGE message.", "err", err)
+				continue
+			}
+			if roundChangeMsg.HasPreparedCertificate() {
+				request = &istanbul.Request{
+					Proposal: roundChangeMsg.PreparedCertificate.Proposal,
+				}
+				break
 			}
-			c.sendPreprepare(r)
-		} else if c.current.pendingRequest != nil {
-			c.sendPreprepare(c.current.pendingRequest)
+		}
+		if request != nil {
+			c.sendPreprepare(request, roundChangeCertificate)
 		}
 	}
 	c.newRoundChangeTimer()
 
 	logger.Debug("New round", "new_round", newView.Round, "new_seq", newView.Sequence, "new_proposer", c.valSet.GetProposer(), "valSet", c.valSet.List(), "size", c.valSet.Size(), "isProposer", c.isProposer())
 }
 
-func (c *core) catchUpRound(view *istanbul.View) {
```

Comment:
I rolled this functionality into `startNewRound` because this would basically cause a round change except it would not update the next proposer which was causing a couple of issues.